src/crimson/os/seastore/transaction_manager.cc

```diff
@@ -557,7 +557,7 @@ TransactionManager::rewrite_extent_ret TransactionManager::rewrite_extent(
   });
 }
 
-TransactionManager::get_extent_if_live_ret TransactionManager::get_extent_if_live(
+TransactionManager::get_extents_if_live_ret TransactionManager::get_extents_if_live(
```

Comment:
IIUC, without splitting pr, it's also possible to make a loophole that there may be some transactions release extents located in reclaim range. With splitting pr, it's impossible to split the same original extent forever and it's a small probability event, so it doesn't make things worse.

Reply:
Oh, I didn't mean that this PR is related to the potential loophole.

Instead, I am trying to evaluate whether we should going forward to change `get_extent_if_live` to `get_extents_if_live`, or fix that hole instead.

Reply:
Hmm, IIRC the original purpose of get_extent_if_live is that the SegmentCleaner didn't get at these extents by reading the lba tree, it got them by scanning the segment headers.  Thus, it still needed to confirm whether the extent in question was still live.  Now, however, we get them via the backref extent.  That really should be transactionally up-to-date and would indeed cause a conflict otherwise.  I think we can actually eliminate this entirely.  @xxhdx1985126 ?

Reply:
> Hmm, IIRC the original purpose of get_extent_if_live is that the SegmentCleaner didn't get at these extents by reading the lba tree, it got them by scanning the segment headers. Thus, it still needed to confirm whether the extent in question was still live. Now, however, we get them via the backref extent. That really should be transactionally up-to-date and would indeed cause a conflict otherwise. I think we can actually eliminate this entirely. @xxhdx1985126 ?

Yes, I think so:-)

However, IIUC, GC always need to read extents out, add them to Cache and then rewrite those extents, and "adding them to Cache" is mandatory for the sake of correctness which is the same reason why, when we prevent GC from touching the LRU, we still have to add extents that are to be GC'd into Cache. Am I right? If so, I think, as long as we have to add extents to Cache, there will be a possibility for GC transactions to conflict with other transactions on those extents, no matter whether we check the liveness of extents or not. Is this right?

With regard to how we should modify the space reclamation procedure, I think, since we have to read extents and rewrite them when reclaiming disk space, we may still have to go through pretty much what `TransactionManager::get_extent_if_live()` is doing now, which is, first try to get extents directly from Cache, and then, if it fails, go through the lba tree to get laddr->paddr mappings and read the extents from disk. The reason why we still have to go through the lba tree is that we have to maintain the invariant that all parent nodes in the lba tree must have already been loaded into Cache before child ones are loaded. Am I right? If so, I think the focus is whether we should maintain that invariant. If we do, I think maybe we should drop `TransactionManager::get_extent_if_live()` and call `TransactionManager::read_extent()` instead. But it will be a totally different situation, if we don't. Am I right? @athanatos What do you think?:-)

Reply:
Ah, so we'd need something like read_extent_maybe_cached(..., laddr_t laddr, paddr_t paddr) which checks the cache for paddr and invokes read_extent if not present.

Reply:
> Ah, so we'd need something like read_extent_maybe_cached(..., laddr_t laddr, paddr_t paddr) which checks the cache for paddr and invokes read_extent with laddr if not present.

Yeah, I'll submit a PR for this:-)

Reply:
I think the crux of this problem is that we will need to get alive extents from Cache's backref buffer and read them out in the same continuation (meaning atomically among other current transactions). Otherwise, if any transactions can be committed in-between, the paddr~len to read out might be already splitted.

IIUC the lba tree is not quite related to this issue as we will always need to access it in order to rewrite the extent.

@xxhdx1985126 @athanatos Do you think so?

Reply:
> IIUC the lba tree is not quite related to this issue as we will always need to access it in order to rewrite the extent.

Yes, I mentioned the lba tree just to discuss how we should change the `get_extent_if_live` part of the space reclamation procedure, not for resolving the conflict issue. Since live extents may be at very low level of the lba tree and we have to make sure parent nodes are loaded before child nodes, we have to go through the lba tree any way.

What I meant in https://github.com/ceph/ceph/pull/45519#discussion_r917218944 are as follows:

1. the checking of liveness of extents in space reclamation is not necessary, indeed;
2. dropping that checking may not help to solve the conflict issue, as we still have to read extents out and add them to Cache even without the liveness checking;
3. we should drop the liveness checking even if it doesn't help to solve the conflict issue, and we have to involve a process pretty similar to  what `TransactionManager::get_extent_if_live()` is doing now but not for the purpose of extent liveness checking

BTW, since the conflict rate is very low now, and, as @zhscn mentioned above, gc transactions won't be locked in an infinite loop of conflict(because extents are at least "block size" and can't be splitted forever), should we put aside this issue for now?

Reply:
Yeah, I'm fine to leave it in a future PR. And this logic can simplified. I'd be happy to do this.